cmds/exp/sed/sed.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2023 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// By Ahmed Kamal <email.ahmedkamal@googlemail.com>
+
+// sed edits file contents using regular expressions.
+//
+// Synopsis:
+//
+//	sed [-ie] [FILE]...
+//
+// Options:
+//
+//	-i, --in-place        in-place file edit
+//	-e, --expression      edits to perform (only s/foo/bar style supported)
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"regexp"
+	"strings"
+
+	flag "github.com/spf13/pflag"
+)
+
+// Writes to a temp file first
+// Renames to the final file upon closing
+type tmpWriter struct {
+	ftmp     os.File
+	filename string
+}
+
+func newTmpWriter(filename string) (*tmpWriter, error) {
+	ftmp, err := os.CreateTemp("/tmp", ".sed*.txt")
+	if err != nil {
+		return nil, fmt.Errorf("unable to create temp file: %w", err)
+	}
+	return &tmpWriter{filename: filename, ftmp: *ftmp}, nil
+}
+
+func (tw *tmpWriter) Write(b []byte) (int, error) {
+	return tw.ftmp.Write(b)
+}
+
+func (tw *tmpWriter) Close() error {
+	err := tw.ftmp.Close()
+	if err != nil {
+		return err
+	}
+	os.Rename(tw.ftmp.Name(), tw.filename)
+	return nil
+}
+
+type transform struct {
+	from   *regexp.Regexp
+	to     string
+	global bool
+}
+
+func (t *transform) run(input io.Reader) io.Reader {
+	pr, pw := io.Pipe()
+	scanner := bufio.NewScanner(input)
+	go func() {
+		defer pw.Close()
+		for scanner.Scan() {
+			line := scanner.Text()
+			var replaced string
+			if t.global {
+				replaced = t.from.ReplaceAllString(line, t.to)
+			} else {
+				counter := 0
+				// Replaces only the first occurrence which golang stdlib does not offer!
+				replaced = t.from.ReplaceAllStringFunc(line, func(value string) string {
+					if counter == 1 {
+						return value
+					}
+					counter++
+					return t.from.ReplaceAllString(value, t.to)
+				})
+			}
+			pw.Write([]byte(replaced))
+			pw.Write([]byte("\n")) // We unconditionally send \n after every line which could be wrong on the last line!
+		}
+	}()
+	return pr
+}
+
+type transforms []transform
+
+func (t *transforms) String() string {
+	return fmt.Sprint(*t)
+}
+
+func (t *transforms) parse(expr []string) (transforms, error) {
+	const MinSearchExprParts = 3
+	const MaxSearchExprParts = 4
+	var ts transforms
+	for _, e := range expr {
+		switch e[:1] {
+		case "s":
+			separator := e[1:2]
+			eParts := strings.Split(e, separator)
```

Comment:
You probably can't use strings.Split here, the problem is escaping.

See this example:

input file: `/home/dir`
sed -e 's/\/home/\/Home/' file
/Home/dir

./sed -e 's/\/home/\/Home/' file
2023/10/03 08:16:44 error parsing expressions: unable to parse transformation. This should be of the form s/old/new/

Also you allow such transformations:
./sed -e 's/hello/world' file
but sed will return an error:
sed -e 's/hello/world' file 
sed: 1: "s/hello/world
": unescaped newline inside substitute pattern